tv/tv.cpp

```diff
@@ -239,4 +262,110 @@ struct TVPass final : public llvm::FunctionPass {
 char TVPass::ID = 0;
 llvm::RegisterPass<TVPass> X("tv", "Translation Validator", false, false);
 
+
+
+#ifdef CLANG_PLUGIN
+/// Classes and functions for running translation validation on clang
+/// Uses new pass manager's callback.
+
+struct TVInitPass : public llvm::PassInfoMixin<TVInitPass> {
+  llvm::PreservedAnalyses run(llvm::Module &M,
+                              llvm::ModuleAnalysisManager &MAM) {
+    TVPass().doInitialization(M);
+    return llvm::PreservedAnalyses::all();
+  }
+};
+
+struct TVFinalizePass : public llvm::PassInfoMixin<TVFinalizePass> {
+  static bool finalized;
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &FAM) {
+    if (!finalized) {
+      finalized = true;
+      TVPass().doFinalization(*F.getParent());
+    }
+    return llvm::PreservedAnalyses::all();
+  }
+};
+
+bool TVFinalizePass::finalized = false;
+
+// Extracting Module out of IR unit.
+// Excerpted from LLVM's StandardInstrumentation.cpp
+static const llvm::Module * unwrapModule(llvm::Any IR) {
+  using namespace llvm;
+
+  if (any_isa<const Module *>(IR))
+    return any_cast<const Module *>(IR);
+  else if (any_isa<const llvm::Function *>(IR))
+    return any_cast<const llvm::Function *>(IR)->getParent();
+  else if (any_isa<const LazyCallGraph::SCC *>(IR)) {
+    auto C = any_cast<const LazyCallGraph::SCC *>(IR);
+    assert(C->begin() != C->end()); // there's at least one function
+    return C->begin()->getFunction().getParent();
+  } else if (any_isa<const Loop *>(IR))
+    return any_cast<const Loop *>(IR)->getHeader()->getParent()->getParent();
+
+  llvm_unreachable("Unknown IR unit");
+}
+
+static bool do_skip(const llvm::StringRef &ref) {
+  const vector<string_view> pass_list = {
+    "::TVInitPass", "::TVFinalizePass",
+    "ArgumentPromotionPass", "DeadArgumentEliminationPass",
+    "HotColdSplittingPass", "InlinerPass",
+    "GlobalOptPass", "IPSCCPPass"
+  };
```

Comment:
I kept this as a vector to use this as `std::find_if(pass_list.begin(), pass_list.end(), ..)` below.